source/grid/manifold_lib.cc

```diff
@@ -281,71 +346,151 @@ get_new_point (const ArrayView<const Point<spacedim>> &vertices,
+    Tensor<1,2> gradlocal;
+
+    while (true)
+      {
+        // Get local basis for estimate c = xVec.:
+        const Tensor<1,3> Clocalx = internal::compute_normal(xVec);
+        const Tensor<1,3> Clocaly = cross_product_3d(xVec, Clocalx);
+
+        // For each vertices vector, compute the tangent vector from xVec
+        // towards the vertices vector -- its length is the spherical length
+        // from xVec to the vertices vector.
+        //Then compute its contribution to the Hessian.
+        gradient = 0.;
+        Hessian = 0.;
+        for ( unsigned int i=0; i<n_points; ++i)
           {
-            normal[0]=-1.;
-            normal[1]=-1.;
-            normal[2]=(vector[0]+vector[1])/vector[2];
+            vPerp = internal::ProjectPerpUnitDiff ( directions[i], xVec );
+            const double sintheta = vPerp.norm();
+            if ( sintheta==0.)
+              {
+                Hessian[0][0]+=weights[i];
+                Hessian[1][1]+=weights[i];
+              }
+            else
+              {
+                const double costheta = (directions[i])*xVec;
+                const double theta = atan2( sintheta, costheta );   // Angle from xVec to (*vi)
+                const double sinthetaInv = 1.0/sintheta;
+
+                vPerp *= sinthetaInv;                       // Normalize
+                const double cosphi = vPerp*Clocalx;
+                const double sinphi = vPerp*Clocaly;
+
+                gradlocal[0] = cosphi;
+                gradlocal[1] = sinphi;
+                gradient += ((weights[i])*theta)*gradlocal; // Added on weighted discrepancy to gradient
+
+                const double sinphiSq = sinphi*sinphi;
+                const double cosphiSq = cosphi*cosphi;
+                const double tt = (weights[i])*(theta*sinthetaInv)*costheta;
+                const double offdiag = cosphi*sinphi*((weights[i])-tt);
+                Hessian[0][0] += (weights[i])*cosphiSq+tt*sinphiSq;
+                Hessian[0][1] += offdiag;
+                Hessian[1][0] += offdiag;
+                Hessian[1][1] += (weights[i])*sinphiSq+tt*cosphiSq;
+              }
           }
-        return normal;
+
+        const double det = Hessian[0][0]*Hessian[1][1]-Hessian[0][1]*Hessian[1][0];
+        Assert(det>1.e-16, ExcInternalError());
+
+        const Tensor<2,2> inverse_Hessian = invert(Hessian);
+
+        const Tensor<1,2> xDisplocal = inverse_Hessian*gradient;
+        const Tensor<1,3> xDisp = xDisplocal[0]*Clocalx + xDisplocal[1]*Clocaly;
+
+//        std::cout << "    xDisp = " << xDisp << "\n";  // DEBUG
+
+        // Step 2c: rotate xVec in direction xDisp, for new estimate.
+
+        const Tensor<1,3> xVecOld = xVec;
+        xVec = internal::RotateUnitInDirection(xVec, xDisp);
+        xVec /= xVec.norm();                // Avoid roundoff error problems
+
+        //      std::cout << xVec << "\n";  // DEBUG
+
+        const double error = (xVec-xVecOld).norm();
+        if ( error <= tolerance )
```

Comment:
use `error = (xVec-xVecOld).norm_square()` here and compare with `tolerance*tolerance` -> saves a square root